lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -1174,6 +1174,111 @@ struct ShrOpConversion : public OpConversionPattern<ShrOp> {
+
+    Value initVal = rewriter.create<hw::ConstantOp>(loc, resultType, 1);
+    Value lhsVal = rewriter.create<ConversionOp>(loc, resultType, op.getLhs());
+
+    auto forOp = rewriter.create<scf::ForOp>(
+        loc, lowerBound, upperBound, step, ValueRange(initVal),
+        [&](OpBuilder &builder, Location loc, Value i, ValueRange iterArgs) {
+          Value loopVar = iterArgs.front();
+          Value mul = rewriter.create<comb::MulOp>(loc, lhsVal, loopVar);
+          rewriter.create<scf::YieldOp>(loc, ValueRange(mul));
+        });
+
+    rewriter.replaceOp(op, forOp.getResult(0));
+
+    return success();
+  }
+};
+
+struct PowSOpConversion : public OpConversionPattern<PowSOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(PowSOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Type resultType = typeConverter->convertType(op.getResult().getType());
+
+    Location loc = op.getLoc();
+    auto intType = cast<IntType>(op.getRhs().getType());
+    if (auto baseOp = op.getLhs().getDefiningOp<ConstantOp>()) {
+      if (baseOp.getValue() == 2) {
+        Value constOne = rewriter.create<ConstantOp>(loc, intType, 1);
+        Value constZero = rewriter.create<ConstantOp>(loc, intType, 0);
+        Value shift = rewriter.create<ShlOp>(loc, constOne, op.getRhs());
+        Value isNegative = rewriter.create<SltOp>(loc, op.getRhs(), constZero);
+        auto condOp = rewriter.replaceOpWithNewOp<ConditionalOp>(
+            op, op.getLhs().getType(), isNegative);
+        Block *thenBlock = rewriter.createBlock(&condOp.getTrueRegion());
+        rewriter.setInsertionPointToStart(thenBlock);
+        rewriter.create<YieldOp>(loc, constZero);
+        Block *elseBlock = rewriter.createBlock(&condOp.getFalseRegion());
+        rewriter.setInsertionPointToStart(elseBlock);
+        rewriter.create<YieldOp>(loc, shift);
+        return success();
+      }
+    }
+
+    // transform a ** b into scf.for 0 to b step 1 { init *= a }, init = 1
+    Type integerType = rewriter.getIntegerType(intType.getWidth());
+    Value lhsVal = rewriter.create<ConversionOp>(loc, resultType, op.getLhs());
+    Value rhsVal = rewriter.create<ConversionOp>(loc, integerType, op.getRhs());
+    Value constZero = rewriter.create<hw::ConstantOp>(loc, integerType, 0);
+    Value isNegative = rewriter.create<comb::ICmpOp>(loc, ICmpPredicate::slt,
+                                                     rhsVal, constZero);
+
+    // if the exponent is negative, lhsVal = 1 / lhs, upperBound = -rhs
+    Value invLhs = rewriter.create<comb::DivUOp>(
+        loc, rewriter.create<hw::ConstantOp>(loc, resultType, 1), lhsVal);
+    Value negRhs = rewriter.create<comb::SubOp>(loc, constZero, rhsVal);
+
+    lhsVal = rewriter.create<comb::MuxOp>(loc, isNegative, invLhs, lhsVal);
+    Value upperBound =
+        rewriter.create<comb::MuxOp>(loc, isNegative, negRhs, rhsVal);
```

Comment:
If the exponent is negative, can this ever produce anything besides 0? Since these are integers, I'd assume `1 / lhs` to always return 0, since there are no fractional numbers. We'll have to figure out a way to represent real numbers, but you could check if the `PowSOp` type is a `moore::IntType`, and only implement integers for now. (We don't really have a lowering for SystemVerilog's `real` type yet. That would be cool to add at some point -- but we'd probably have to map that to the `arith` MLIR dialect and MLIR's standard floating-point type.)

Reply:
This is a really good point, and I hadn't considered this question before. Is there a better way to handle the negative exponent situation? Should I remove the `comb::DivUOp` and skip the `scf.for` loop, then return const zero like the `PowSOp::canonicalize` function did?

Reply:
Thank you so much for your guidance and patience!  Based on your suggestions, I've made some updates to the code and tests. 😃